src/JobManager.cpp

```diff
@@ -0,0 +1,449 @@
+        this->thread_info[thread->get_id()] = std::tuple<int, int, ThreadPool::function_type, std::shared_ptr<Job>>{i, count, func, job};
+    }
+
+    this->unlock();
+    return *out;
+}
+
+std::vector<std::shared_ptr<std::thread>> &ThreadPool::request_upto(unsigned int count, ThreadPool::function_type func,
+                                                                   std::shared_ptr<Job> &job) {
+    this->lock();
+    auto *out = new std::vector<std::shared_ptr<std::thread>>();
+
+    int total = 0;
+
+    for (int i = 0; i < count; i++) {
+        // Check that there are threads available.
+        if (this->avail.size() == 0) {
+            break;
+        }
+        // Get the new thread.
+        std::shared_ptr<std::thread> thread = this->avail.back();
+        this->running.push_back(thread);
+        out->push_back(thread); 
+        this->avail.pop_back(); 
+        total++;
+    }
+
+    // Push the thread info.
+    for (int i = 0; i < total; i++) {
+        this->thread_info[out->at(i)->get_id()] = std::tuple<int, int, ThreadPool::function_type, std::shared_ptr<Job>>{i, total, func, job};
+    }
+
+    this->unlock();
+    return *out;
+}
+
+void ThreadPool::release(std::vector<std::shared_ptr<std::thread>> &threads) {
+    this->lock();
+
+    // Move running threads back to the waiting pool.
+    for(auto thread : threads) {
+      for(size_t i = 0; i < this->running.size(); i++) {
+	if(this->running[i]->get_id() == thread->get_id()) {
+	  this->avail.push_back(this->running[i]);
+	  this->thread_info[this->running[i]->get_id()] =
+	    std::tuple<int, int, ThreadPool::function_type, std::shared_ptr<Job>>{0, 0, nullptr, nullptr};
+	  this->running.erase(std::next(this->running.begin(), i));
+	  i--;
+	}
+      }
+    }
+
+    this->unlock();
+}
+
+int ThreadPool::index(std::thread::id id) {
+    this->lock();
+
+    int out = std::get<0>(this->thread_info[id]);
+
+    this->unlock();
+    return out;
+}
+
+int ThreadPool::compute_threads(std::thread::id id) {
+    this->lock();
+
+    int out = std::get<1>(this->thread_info[id]);
+
+    this->unlock();
+    return out;
+}
+
+ThreadPool::function_type ThreadPool::compute_function(std::thread::id id) {
+    this->lock();
+
+    ThreadPool::function_type out = std::get<2>(this->thread_info[id]);
+
+    this->unlock();
+    return out;
+}
+
+std::shared_ptr<Job> ThreadPool::thread_job(std::thread::id id) {
+    this->lock();
+
+    std::shared_ptr<Job> out = std::get<3>(this->thread_info[id]);
+
+    this->unlock();
+    return out;
+}
+
+void ThreadPool::thread_finished(std::thread::id id) {
+    this->lock();
+
+    this->thread_info[id] = std::tuple<int, int, ThreadPool::function_type, std::shared_ptr<Job>>{0, 0, nullptr, nullptr};
+
+    for (size_t i = 0; i < this->running.size(); i++) {
+        if (this->running[i]->get_id() == id) {
+            this->avail.push_back(this->running[i]);
+            this->running.erase(std::next(this->running.begin(), i));
```

Comment:
I may need someone to take a look at this. I just can't seem to find the issue. The code is throwing a segfault here. It seems to me that a stack-allocated instance of shared_ptr is being passed around, which doesn't convey any information about the contents of the shared_ptr to children. I just can't seem to find the issue. It only happens in multithreaded mode, so valgrind is less than helpful.

Reply:
I added my vscode files if anyone is interested in trying to replicate. I will remove them when they are no longer needed.

Reply:
Do I have to do anything to make it use more threads or is it automatic?

Reply:
I occasionally get a run that hangs here:

```
    frame #4: 0x000000010017d554 test-all`CATCH2_INTERNAL_TEST_115() [inlined] einsums::jobs::ReadLock<einsums::Tensor<double, 2ul>>::get(this=0x00006000031902b8) at Jobs.hpp:117:13 [opt]
    frame #5: 0x000000010017d540 test-all`CATCH2_INTERNAL_TEST_115() at TensorAlgebra_Jobs.cpp:1977:17 [opt]
```

Reply:
I don't know if it's necessarily important to note that this is hanging on a yield statement. Still working on fixing the bugs, but could it be something with different schedulers? (probably not, probably a bug.)